README.md
# Parse

Detect and convert short strings into integers, floats, dates, times, booleans, arrays, and hashes - "like a human would".

## Note on versions

You can always use `Parse.parse`. It will always point to the most recent version of the algorithm (currently `Parse.ver0_1_0`).

If the algorithm changes and you need the old version, you can reference it by its version number. For example, `Parse.ver0_0_1`.

Since almost any change to the algorithm is a breaking change, there are going to be lots of minor version bumps (as opposed to patches).

## Usage

You get the idea:

    Parse.parse("15,000")                        #=> 15000
    Parse.parse("$123.4")                        #=> 123.4
    Parse.parse("(10,000,000.00)")               #=> -10000000.0
    Parse.parse("true")                          #=> true
    Parse.parse("no")                            #=> true
    Parse.parse("1982-01-01")                    #=> Fri, 01 Jan 1982
    Parse.parse("2010-05-05 13:42:16 Z")         #=> 2010-05-05 10:42:16 -0300
    Parse.parse("2010-05-05 13:42:16 -02:00")    #=> 2010-05-05 12:42:16 -0300
    Parse.parse("100%")                          #=> 1.0
    Parse.parse("50%")                           #=> 0.5
    Parse.parse("12/25/82", date: :us)           #=> Sat, 25 Dec 1982
    Parse.parse("25/12/82", date: :euro)         #=> Sat, 25 Dec 1982
    Parse.parse("#DIV/0")                        #=> Infinity
    Parse.parse("")                              #=> nil
    Parse.parse("nil")                           #=> nil
    Parse.parse("#NAME?")                        #=> nil
    Parse.parse("NaN")                           #=> NaN
    Parse.parse(".NaN")                          #=> NaN
    Parse.parse("-.inf")                         #=> -Infinity
    Parse.parse("Inf")                           #=> "Inf"
    Parse.parse(":no_symbols")                   #=> ":no_symbols"

More esoteric stuff:

    Parse.parse("-")                             #=> nil
    Parse.parse("?")                             #=> nil
    Parse.parse("-8e-05")                        #=> -8.0e-05
    Parse.parse("-12.5e-13")                   #=> -1.25e-12
    Parse.parse("05753")                         #=> 5753
    Parse.parse("15000")                        #=> 15000
    Parse.parse("15.0")                          #=> 15.0
    Parse.parse("15,000.0")                      #=> 15000.0
    Parse.parse("15000.0")                      #=> 15000.0
    Parse.parse("0015")                          #=> 15
    Parse.parse("0015.0")                        #=> 15.0
    Parse.parse("0015.0")                       #=> 15.0
    Parse.parse("0x15")                          #=> 21
    Parse.parse("0o15")                          #=> 13
    Parse.parse("8e-05")                         #=> 8.0e-05
    Parse.parse("12.5e-13")                    #=> 1.25e-12
    Parse.parse("0$123.4")                       #=> 123.4
    Parse.parse("$15,000")                       #=> 15000
    Parse.parse("0$15,000")                      #=> 15000
    Parse.parse("$123456")                      #=> 123456
    Parse.parse("$123456.7")                    #=> 123456.7
    Parse.parse("10,000,000")                    #=> 10000000
    Parse.parse("10,000,000.00")                 #=> 10000000.0
    Parse.parse("$10,000,000.00")                #=> 10000000.0
    Parse.parse("0$10,000,000.00")               #=> 10000000.0
    Parse.parse("$010,000,000.00")               #=> 10000000.0
    Parse.parse("-15")                           #=> -15
    Parse.parse("-15,000")                       #=> -15000
    Parse.parse("-15000")                       #=> -15000
    Parse.parse("-15.0")                         #=> -15.0
    Parse.parse("-15,000.0")                     #=> -15000.0
    Parse.parse("-15000.0")                     #=> -15000.0
    Parse.parse("-15000.0")                    #=> -15000.0
    Parse.parse("00-15")                         #=> -15
    Parse.parse("00-15.0")                       #=> -15.0
    Parse.parse("-0x15")                         #=> -21
    Parse.parse("-0o15")                         #=> -13
    Parse.parse("-$123.4")                       #=> -123.4
    Parse.parse("($123.4)")                      #=> -123.4
    Parse.parse("0($123.4)")                     #=> -123.4
    Parse.parse("-$15,000")                      #=> -15000
    Parse.parse("($15,000)")                     #=> -15000
    Parse.parse("-$123,456")                     #=> -123456
    Parse.parse("($123,456)")                    #=> -123456
    Parse.parse("-$123,456.7")                   #=> -123456.7
    Parse.parse("($123,456.7)")                  #=> -123456.7
    Parse.parse("-10,000,000")                   #=> -10000000
    Parse.parse("(10,000,000)")                  #=> -10000000
    Parse.parse("-10,000,000.00")                #=> -10000000.0
    Parse.parse("(10,000,000.00)")               #=> -10000000.0
    Parse.parse("1,200")                         #=> 1200
    Parse.parse("1,200.0")                       #=> 1200.0
    Parse.parse("1,2")                           #=> 12
    Parse.parse("1,20")                          #=> 120
    Parse.parse("1,2.0")                         #=> 12.0
    Parse.parse("1.0,2")                         #=> "1.0,2"
    Parse.parse("1.0,2.0")                       #=> "1.0,2.0"
    Parse.parse("-1,200")                        #=> -1200
    Parse.parse("-1,200.0")                      #=> -1200.0
    Parse.parse("-1,2")                          #=> -12
    Parse.parse("-1,20")                         #=> -120
    Parse.parse("-1,2.0")                        #=> -12.0
    Parse.parse("-1.0,2")                        #=> "-1.0,2"
    Parse.parse("-1.0,2.0")                      #=> "-1.0,2.0"
    Parse.parse("01,200")                        #=> 1200
    Parse.parse("01,200.0")                      #=> 1200.0
    Parse.parse("01,2")                          #=> 12
    Parse.parse("01,20")                         #=> 120
    Parse.parse("01,2.0")                        #=> 12.0
    Parse.parse("01.0,2")                        #=> "01.0,2"
    Parse.parse("01.0,2.0")                      #=> "01.0,2.0"
    Parse.parse("#hello")                        #=> "#hello"
    Parse.parse("\n#hello\n#world")              #=> "#hello #world"
    Parse.parse("hello\nworld")                  #=> "hello world"
    Parse.parse(",1")                            #=> ",1"
    Parse.parse(",1,")                           #=> ",1,"

Note how, for better or worse, it effectively acts as a YAML or JSON parser, but doesn't do CSV:

    Parse.parse("1,2,3")                         #=> "1,2,3"
    Parse.parse("[1,2,3]")                       #=> [1, 2, 3]
    Parse.parse("---\na: 1\n")                   #=> {"a"=>1}
    Parse.parse("---\n:a: 1\n")                  #=> {":a"=>1}
    Parse.parse("---\na: 1\n5: |-\n  c\n  3\n")  #=> {"a"=>1, 5=>"c\n3"}
    Parse.parse("{\"a\":1}")                     #=> {"a"=>1}
    Parse.parse("{\"a\":1,\"5\":\"c\\n3\"}")     #=> {"a"=>1, "5"=>"c\n3"}

## Wishlist

1. allow specifying `date: '%Y/%m/%d'` for strptime
1. allow hinting like `type: Integer` or `type: Date`
1. deprecate this whole thing in favor of YAML 1.2?

## Copyright

2014 Seamus Abshere


CHANGELOG
0.1.0

* Breaking changes

  * Not parsed as numbers anymore: "10_14" or "10_400" or "-1_2e3_5" or "-$123_456.7" -- very unlikely to see in the wild
  * Not parsed as numbers anymore: '1,2' or '-1,2' or '1,20' or '1,22.0' -- too hard to distinguish from CSV

* Enhancements

  * Using proper versioning scheme - every version of algorithm involves a minor bump!
  * Recover from YAML parsing errors in 1.9

0.0.1.1 / 2014-02-06

* Bug fixes

  * Parse.parse called Parse.ver0 instead of Parse.ver0_0_1. Doh! 

0.0.1 / 2014-02-06 [yanked]

Initial release!


spec/parse_spec.rb
require 'spec_helper'

describe Parse do
  it "should parse with version 0.0.2 of the algorithm" do
    v = "  1990-04-03  "
    expect(Parse.parse(v)).to eq(Parse.ver0_1_0(v))
  end
end


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'parse'

require 'pry'
require 'multi_json'
require 'active_support/core_ext'


spec/parse_ver0_0_1_spec.rb
require 'spec_helper'

describe Parse do
  {
    ''                                 => nil,
    'nil'                              => nil,
    '15'                               => 15,
    '15,000'                           => 15_000,
    '15_000'                           => 15_000,
    '15_00_0'                          => 15_000,
    '15.0'                             => 15.0,
    '15,000.0'                         => 15_000.0,
    '15_000.0'                         => 15_000.0,
    '15_00_0.0'                        => 15_000.0,
    '0015'                             => 15,   # not octal
    '0015.0'                           => 15.0,  # not octal
    '0_015.0'                          => 15.0, # just weird
    '0x15'                             => 0x15, # hex
    '0o15'                             => 015,  # octal
    '8e-05'                            => 8e-05,
    '1_2.5e-1_3'                       => 12.5e-13,
    '$123.4'                           => 123.4,
    '0$123.4'                           => 123.4,
    '$15,000'                          => 15_000,
    '0$15,000'                          => 15_000,
    '$123_456'                         => 123_456,
    '$123_456.7'                       => 123_456.7,
    '10,000,000'                       => 10_000_000,
    '10,000,000.00'                    => 10_000_000.0,
    '$10,000,000.00'                    => 10_000_000.0,
    '0$10,000,000.00'                    => 10_000_000.0,
    '$010,000,000.00'                    => 10_000_000.0,

    '-15'                              => -15,
    '-15,000'                          => -15_000,
    '-15_000'                          => -15_000,
    '-15_00_0'                         => -15_000,
    '-15.0'                            => -15.0,
    '-15,000.0'                        => -15_000.0,
    '-15_000.0'                        => -15_000.0,
    '-15_00_0.0'                       => -15_000.0,
    '00-15'                            => -15,   # not octal
    '00-15.0'                          => -15.0,  # not octal
    '0_0-15.0'                         => '0_0-15.0', # just weird
    '-0x15'                            => -0x15, # hex
    '-0o15'                            => -015,  # octal
    '-8e-05'                           => -8e-05,
    '-1_2.5e-1_3'                      => -12.5e-13,
    '-$123.4'                          => -123.4,
    '($123.4)'                         => -123.4,
    '0($123.4)'                        => -123.4,
    '-$15,000'                         => -15_000,
    '($15,000)'                        => -15_000,
    '-$123_456'                        => -123_456,
    '($123_456)'                       => -123_456,
    '-$123_456.7'                      => -123_456.7,
    '($123_456.7)'                     => -123_456.7,
    '-10,000,000'                      => -10_000_000,
    '(10,000,000)'                     => -10_000_000,
    '-10,000,000.00'                   => -10_000_000.0,
    '(10,000,000.00)'                  => -10_000_000.0,
    '-10000000'                        => -10_000_000,
    '(10000000)'                       => -10_000_000,
    '-10000000.00'                     => -10_000_000.0,
    '(10000000.00)'                    => -10_000_000.0,
    '1,200'                            => 1_200,
    '1,200.0'                          => 1_200.0,
    '1,2'                              => 12,
    '1,20'                             => 120,
    '1,2.0'                            => 12.0,
    '1.0,2'                            => '1.0,2',
    '1.0,2.0'                          => '1.0,2.0',
    '-1,200'                           => -1_200,
    '-1,200.0'                         => -1_200.0,
    '-1,2'                             => -12,
    '-1,20'                            => -120,
    '-1,2.0'                           => -12.0,
    '-1.0,2'                           => '-1.0,2',
    '-1.0,2.0'                         => '-1.0,2.0',
    '01,200'                           => 1_200,
    '01,200.0'                         => 1_200.0,
    '01,2'                             => 12,
    '01,20'                            => 120,
    '01,2.0'                           => 12.0,
    '01.0,2'                           => '01.0,2',
    '01.0,2.0'                         => '01.0,2.0',

    '05753'                            => 5753,
    'true'                             => true,
    'yes'                              => true,
    'false'                            => false,
    'no'                               => false,
    '#DIV/0'                           => (1.0/0),
    '#NAME?'                           => nil,
    'Inf'                              => 'Inf',
    'Infinity'                         => (1.0/0),
    '-Infinity'                        => -(1.0/0),
    'NaN'                              => 0.0/0, # need the dot
    '.NaN'                             => 0.0/0,  # NaN
    '-.inf'                            => -(1.0/0), # -Infinity
    '-'                                => nil, # per bigml
    '?'                                => nil,
    '1982-01-01'                       => Date.new(1982,1,1),
    '2010-05-05 13:42:16 Z'            => Time.parse('2010-05-05 13:42:16 Z'),
    '2010-05-05 13:42:16 -02:00'       => Time.parse('2010-05-05 13:42:16 -02:00'),
    ":not_a_symbol"                    => ':not_a_symbol',
    '#hello'                           => '#hello',
    "\n#hello\n#world"                 => '#hello #world',
    "hello\nworld"                     => 'hello world', # whitespace compression

    '0%'                               => 0.0,
    '100%'                             => 1.0,
    '50%'                              => 0.5,
    '5%'                               => 0.05,
    '00000%'                           => 0.0,
    '0000100%'                         => 1.0,
    '000050%'                          => 0.5,
    '00005%'                           => 0.05,

    ['12/25/82', {date: :us}]          => Date.new(1982,12,25),
    ['12/25/1982', {date: :us}]        => Date.new(1982,12,25),
    ['25/12/82', {date: :euro}]        => Date.new(1982,12,25),
    ['25/12/1982', {date: :euro}]      => Date.new(1982,12,25),
    ['12-25-82', {date: :us}]          => Date.new(1982,12,25),
    ['12-25-1982', {date: :us}]        => Date.new(1982,12,25),
    ['25-12-82', {date: :euro}]        => Date.new(1982,12,25),
    ['25-12-1982', {date: :euro}]      => Date.new(1982,12,25),

    '12/25/82'                         => '12/25/82',

    ',1'                               => ',1', # not a csv parser
    ',1,'                              => ',1,', # not a csv parser
    '1,2,3'                            => '1,2,3', # not a csv parser
    '[1,2,3]'                          => [1,2,3],
    YAML.dump('a' => 1)                => { 'a' => 1 },
    YAML.dump(a: 1)                    => { ':a' => 1 }, # doesn't parse symbols
    YAML.dump('a' => 1, 5 => "c\n3")   => { 'a' => 1, 5 => "c\n3" },
    MultiJson.dump(a: 1)               => { 'a' => 1 }, # json always loses symbols
    MultiJson.dump(a: 1, 5 => "c\n3")  => { 'a' => 1, '5' => "c\n3" },
  }.each do |input, expected|
    it "parses #{input.inspect} as #{expected.inspect}" do
      input = Array.wrap input
      got = Parse.ver0_0_1(*input)
      # $lines << [ "Parse.parse(#{input.inspect})".ljust(45), "#=> #{got.inspect}" ].join
      if expected.is_a?(Float) and expected.nan?
        expect(got.nan?).to eq(true)
      else
        expect(got).to eq(expected)
      end

      input_with_spaces = [ "\t" + input[0] + "\t", input[1] ]
      got_with_spaces = Parse.ver0_0_1(*input_with_spaces)
      if expected.is_a?(Float) and expected.nan?
        expect(got.nan?).to eq(true)
      else
        expect(got_with_spaces).to eq(expected)
      end
    end
  end
end


lib/parse/algorithm/ver0_0_1.rb
module Parse
  module Algorithm
    class Ver0_0_1
      NULL = [ '', '-', '?', 'N/A', 'n/a', 'NULL', 'null', '#REF!', '#NAME?', 'NIL', 'nil', 'NA', 'na', '#VALUE!', '#NULL!'] # from bigml's list
      DATE = {
        euro: ['%d-%m-%Y', '%d-%m-%y'],
        us:   ['%m-%d-%Y', '%m-%d-%y'],
      }

      attr_reader :raw
      attr_reader :options
      def initialize(raw, options = nil)
        @raw = raw
        @options = options
      end

      # @private
      # use YAML to parse stuff like '1.5'
      # ruby's yaml is 1.1, which means it does weird stuff with '001' (fixed in 1.2, which jruby has)
      def result
        return raw unless raw.is_a? String
        
        memo = raw.strip
      
        return nil if NULL.include? memo
        return 1.0/0 if INFINITY.include? memo
        return -1.0/0 if NEG_INFINITY.include? memo
        return 0.0/0 if NAN.include? memo
        
        if options and options[:date]
          yyyy, yy = DATE.fetch options[:date]
          memo.sub!(/0+/, '')
          memo.gsub! '/', '-'
          if memo =~ /\d{4,}/ # yyyy
            return Date.strptime(memo, yyyy)
          else
            return Date.strptime(memo, yy)
          end
        end

        not_numeric = nil
        not_numeric ||= memo =~ /,\d{1,2},/ # comma not used for thousands, like 10,20,30
        not_numeric ||= memo =~ /\..*,/ # comma following a period, like 1.0,2
        not_numeric ||= memo =~ /\A[^(+\-\$0-9%]/ # starts with letter or smth
        possible_numeric = !not_numeric
        accounting_negative = nil
        percentage = nil

        if possible_numeric
          accounting_negative = memo =~ /\A[0$]*\([0$]*/
          percentage = memo.end_with?('%')
          memo.sub! /%\z/, '' if percentage
          memo.delete!('()') if accounting_negative # accounting negative
          # in yaml 1.1, anything starting with zero is treated as octal... in 1.2, it's 0o
          memo.sub!(/0+/, '') if memo =~ /\A[+\-]?0+[+\-\$]?[1-9]+/ # leading zeros
          memo.delete!('$') if memo =~ /\A[+\-]?0*\$/
          if memo.include?(',')
            a, b = memo.split('.', 2)
            a.delete! ','
            memo = b ? [a, b].join('.') : a
          end
        end

        not_safe_for_yaml = nil
        not_safe_for_yaml ||= memo.include?('#')
        not_safe_for_yaml ||= not_numeric && memo =~ /\A[\d,]+\z/ #1,2,3, maybe a csv
        safe_for_yaml = !not_safe_for_yaml

        if safe_for_yaml
          begin
            memo = SafeYAML.load memo
          rescue
            $stderr.puts "#{memo.inspect} => #{$!}"
          end
        end

        if possible_numeric
          case memo
          when /\A[+\-]?[\d._]+[eE][+\-]?[\d._]+\z/
            # scientific notation
            memo = memo.to_f
          when /\A[+\-]?0o/
            # octal per yaml 1.2
            memo = memo.to_i 8
          end
        end
        
        if memo.is_a?(String)
          # compress whitespace
          memo.gsub! /\s+/, ' '
        end

        memo = memo / 100.0 if percentage
        memo = -memo if accounting_negative
        memo
      end
    end
  end
end


lib/parse/algorithm/ver0_0_2.rb
module Parse
  module Algorithm
    class Ver0_0_2
      NULL = [ '', '-', '?', 'N/A', 'n/a', 'NULL', 'null', '#REF!', '#NAME?', 'NIL', 'nil', 'NA', 'na', '#VALUE!', '#NULL!', '00/00/00', '0000-00-00'] # from bigml's list
      REGION_DATE_FORMAT = {
        euro: ['%d-%m-%Y', '%d-%m-%y'],
        us:   ['%m-%d-%Y', '%m-%d-%y'],
        iso:  ['%Y-%m-%d', '%y-%m-%d'], # second one is silly
      }
      DATE_DETECT = {
        %r{\A0*[12]\d\d\d[\-/](?:(?:0[1-9])|(?:1[0-2]))[\-/][1-9]\d\z} => :iso, # $1 will be delimiter
      }
      EMPTY_OPTIONS = {}

      attr_reader :raw
      attr_reader :options
      def initialize(raw, options = nil)
        @raw = raw
        @options = options || EMPTY_OPTIONS
      end

      def result
        return raw unless raw.is_a? String
        
        memo = raw.strip
      
        return nil if NULL.include? memo
        return 1.0/0 if INFINITY.include? memo
        return -1.0/0 if NEG_INFINITY.include? memo
        return 0.0/0 if NAN.include? memo

        date_region = if options[:date]
          options[:date]
        else
          catch :hit do
            DATE_DETECT.each do |pattern, date_region|
              # binding.pry if memo.include?('2011-')
              if memo =~ pattern
                throw :hit, date_region
              end
            end
            nil
          end
        end

        if date_region.nil? and options[:type] == Date
          date_region = :iso
        end

        if date_region
          yyyy, yy = REGION_DATE_FORMAT.fetch date_region
          is_yyyy = memo =~ /[1-9]\d\d\d/
          memo.sub! /\A0+/, ''
          memo.gsub! '/', '-'
          if is_yyyy
            if memo.length < 10 and date_region == :iso
              return Date.parse(memo)
            else
              return Date.strptime(memo, yyyy)
            end
          else
            return Date.strptime(memo, yy)
          end
        end

        possible_numeric = nil
        not_numeric = nil
        certain_numeric = nil
        if [Numeric, Integer, Float].include?(options[:type])
          certain_numeric = true
          possible_numeric = true
          not_numeric = false
        else
          not_numeric ||= memo =~ /[1-9][^)\d_,%.eE]/ # has a dash in the middle
          not_numeric ||= memo =~ /,\d{1,2},/ # comma not used for thousands, like 10,20,30
          not_numeric ||= memo =~ /\..*,/ # comma following a period, like 1.0,2
          not_numeric ||= memo.scan(/[^\d_,%.eE]/).length > memo.scan(/[\d_,%.eE]/).length
          not_numeric ||= memo =~ /\A[^(+\-\$0-9%]/ # starts with letter or smth
          possible_numeric = !not_numeric
        end
        accounting_negative = nil
        percentage = nil

        if possible_numeric
          accounting_negative = memo =~ /\A[0$]*\([0$]*/
          percentage = memo.end_with?('%')
          memo.sub! /%\z/, '' if percentage
          memo.delete!('()') if accounting_negative # accounting negative
          # in yaml 1.1, anything starting with zero is treated as octal... in 1.2, it's 0o
          memo.sub!(/0+/, '') if memo =~ /\A[+\-]?0+[+\-\$]?[1-9]+/ # leading zeros
          memo.delete!('$') if memo =~ /\A[+\-]?0*\$/
          memo.sub!('D', 'e') if memo =~ /\A[+\-]?[\d.]*\dD[+\-]?[\d.]+\z/ # fortran double precision
          if memo.include?(',')
            a, b = memo.split('.', 2)
            a.delete! ','
            memo = b ? [a, b].join('.') : a
          end
        end

        # binding.pry if memo =~ /sqft/i
        if certain_numeric
          memo.gsub! /[a-z]/i, ''
        end

        not_safe_for_yaml = nil
        not_safe_for_yaml ||= memo.include?('#')
        not_safe_for_yaml ||= memo =~ /\A[@,]/
        not_safe_for_yaml ||= not_numeric && memo =~ /\A[\d,]+\z/ #1,2,3, maybe a csv
        safe_for_yaml = !not_safe_for_yaml

        if safe_for_yaml
          begin
            memo = SafeYAML.load memo
          rescue Exception # Psych::SyntaxError will blow up plain rescue in 1.9.3
            $stderr.puts "#{memo.inspect} => #{$!}"
          end
        end

        if possible_numeric
          case memo
          when /\A[+\-]?[\d._]+[eE][+\-]?[\d._]+\z/
            # scientific notation
            memo = memo.to_f
          when /\A[+\-]?0o/
            # octal per yaml 1.2
            memo = memo.to_i 8
          end
        end
        
        if memo.is_a?(String)
          # compress whitespace
          memo.gsub! /\s+/, ' '
        end

        memo = memo / 100.0 if percentage
        memo = -memo if accounting_negative
        memo
      rescue
        if options and options[:ignore_error]
          # nothing to see here
        else
          raise "#{memo.inspect} => #{$!}"
        end
      end
    end
  end
end


lib/parse/algorithm/ver0_1_0.rb
module Parse
  module Algorithm
    class Ver0_1_0
      NULL = [ '', '-', '?', 'N/A', 'n/a', 'NULL', 'null', '#REF!', '#NAME?', 'NIL', 'nil', 'NA', 'na', '#VALUE!', '#NULL!', '00/00/00', '0000-00-00'] # from bigml's list
      REGION_DATE_FORMAT = {
        euro: ['%d-%m-%Y', '%d-%m-%y'],
        us:   ['%m-%d-%Y', '%m-%d-%y'],
        iso:  ['%Y-%m-%d', '%y-%m-%d'], # second one is silly
      }
      DATE_DETECT = {
        %r{\A0*[12]\d\d\d[\-/](?:(?:0[1-9])|(?:1[0-2]))[\-/][1-9]\d\z} => :iso, # $1 will be delimiter
      }
      EMPTY_OPTIONS = {}

      attr_reader :raw
      attr_reader :options
      def initialize(raw, options = nil)
        @raw = raw
        @options = options || EMPTY_OPTIONS
      end

      def result
        return raw unless raw.is_a? String
        
        memo = raw.strip
      
        return nil if NULL.include? memo
        return 1.0/0 if INFINITY.include? memo
        return -1.0/0 if NEG_INFINITY.include? memo
        return 0.0/0 if NAN.include? memo

        date_region = if options[:date]
          options[:date]
        else
          catch :hit do
            DATE_DETECT.each do |pattern, date_region|
              # binding.pry if memo.include?('2011-')
              if memo =~ pattern
                throw :hit, date_region
              end
            end
            nil
          end
        end

        if date_region.nil? and options[:type] == Date
          date_region = :iso
        end

        if date_region
          yyyy, yy = REGION_DATE_FORMAT.fetch date_region
          is_yyyy = memo =~ /[1-9]\d\d\d/
          memo.sub! /\A0+/, ''
          memo.gsub! '/', '-'
          if is_yyyy
            if memo.length < 10 and date_region == :iso
              return Date.parse(memo)
            else
              return Date.strptime(memo, yyyy)
            end
          else
            return Date.strptime(memo, yy)
          end
        end

        possible_numeric = nil
        not_numeric = nil
        certain_numeric = nil
        if [Numeric, Integer, Float].include?(options[:type])
          certain_numeric = true
          possible_numeric = true
          not_numeric = false
        else
          # not_numeric ||= memo =~ /[1-9][^)\d_,%.eE]/ # has a dash in the middle
          not_numeric ||= memo.include?('_')
          not_numeric ||= memo =~ %r{[1-9][/-]\d}
          not_numeric ||= memo =~ /,\d{1,2}(?:[.\D]|\z)/
          not_numeric ||= memo.scan(/[^\d_,%.eE]/).length > memo.scan(/[\d_,%.eE]/).length
          not_numeric ||= memo =~ /\A[^(+\-\$0-9%]/ # starts with letter or smth
          possible_numeric = !not_numeric
        end
        accounting_negative = nil
        percentage = nil
        if possible_numeric
          accounting_negative = memo =~ /\A[0$]*\([0$]*/
          percentage = memo.end_with?('%')
          memo.sub! /%\z/, '' if percentage
          memo.delete!('()') if accounting_negative # accounting negative
          # in yaml 1.1, anything starting with zero is treated as octal... in 1.2, it's 0o
          memo.sub!(/0+/, '') if memo =~ /\A[+\-]?0+[+\-\$]?[1-9]+/ # leading zeros
          memo.delete!('$') if memo =~ /\A[+\-]?0*\$/
          memo.sub!('D', 'e') if memo =~ /\A[+\-]?[\d.]+D[+\-]?[\d.]+\z/ # fortran double precision
          if memo.include?(',')
            a, b = memo.split('.', 2)
            a.delete! ','
            memo = b ? [a, b].join('.') : a
          end
        end

        if certain_numeric
          memo.gsub! /[a-z]/i, ''
        end

        not_safe_for_yaml = nil
        not_safe_for_yaml ||= memo =~ /\A(on|off)\z/i
        not_safe_for_yaml ||= memo.include?('#')
        not_safe_for_yaml ||= memo =~ /\A[@&,]/
        not_safe_for_yaml ||= not_numeric && memo.start_with?('0')
        not_safe_for_yaml ||= not_numeric && memo =~ /\A[^{\[]*\d[,_]/ #1,2,3, maybe a csv

        safe_for_yaml = !not_safe_for_yaml

        if safe_for_yaml
          begin
            memo = SafeYAML.load memo
          rescue Exception # Psych::SyntaxError will blow up plain rescue in 1.9.3
            $stderr.puts "#{memo.inspect} => #{$!}"
          end
        end

        if possible_numeric
          case memo
          when /\A[+\-]?[\d._]+[eE][+\-]?[\d._]+\z/
            # scientific notation
            memo = memo.to_f
          when /\A[+\-]?0o/
            # octal per yaml 1.2
            memo = memo.to_i 8
          end
        end
        
        if memo.is_a?(String)
          # compress whitespace
          memo.gsub! /\s+/, ' '
        end

        memo = memo / 100.0 if percentage
        memo = -memo if accounting_negative
        memo
      rescue
        if options and options[:ignore_error]
          # nothing to see here
        else
          raise "#{memo.inspect} => #{$!}"
        end
      end
    end
  end
end
